src/variable_rate.jl

```diff
@@ -0,0 +1,397 @@
+        sum_rate = add_fast(new_rate, prev_rate)
+        cur_rates[idx] = sum_rate
+        return total_variable_rate(vjumps, u, p, t, cur_rates, idx + 1, sum_rate)
+    end
+end
+
+# Condition functor defined directly on the cache
+function (cache::VRDirectCBEventCache)(u, t, integrator)
+    if integrator.t != cache.current_time
+        cache.prev_time = cache.current_time
+        cache.prev_threshold = cache.current_threshold
+        cache.current_time = integrator.t
+    end
+    
+    dt = t - cache.prev_time
+    if dt == 0
+        return cache.prev_threshold
+    end
+
+    vjumps = cache.variable_jumps
+    p = integrator.p
+    n = 4
+    rate_increment = zero(t)
+    for i in 1:n
+        τ = ((dt / 2) * gauss_points[n][i]) + ((t + cache.prev_time) / 2)
+        u_τ = integrator(τ)
+        total_variable_rate_τ = total_variable_rate(vjumps, u_τ, p, τ, cache.cur_rates)
+        rate_increment += gauss_weights[n][i] * total_variable_rate_τ
+    end
+    rate_increment *= (dt / 2)
+    
+    cache.current_threshold = cache.prev_threshold - rate_increment
+    
+    return cache.current_threshold
+end
+
+# Affect functor defined directly on the cache
+function (cache::VRDirectCBEventCache)(integrator)
+    t = integrator.t
+    u = integrator.u
+    p = integrator.p
+    rng = cache.rng
+
+    cache.total_rate_cache = total_variable_rate(cache.variable_jumps, u, p, t, cache.cur_rates)
+    total_variable_rate_sum = cache.total_rate_cache
+    if total_variable_rate_sum <= 0
+        return
+    end
+
+    r = rand(rng) * total_variable_rate_sum
+    vjumps = cache.variable_jumps
+    if !isempty(vjumps)
+        @inbounds jump_idx = searchsortedfirst(cache.cur_rates, r)
+        execute_affect!(vjumps, integrator, jump_idx)
+    end
+
+    cache.prev_time = t
+    cache.current_threshold = randexp(rng)
+    cache.prev_threshold = cache.current_threshold
+    cache.current_time = t
+    return nothing
+end
+
+function execute_affect!(vjumps::Tuple{Vararg{VariableRateJump}}, integrator, idx)
+    if !(1 <= idx <= length(vjumps))
+        error("Jump index $idx out of bounds for $(length(vjumps)) jumps")
+    end
+    @inbounds vjumps[idx].affect!(integrator)
+end
+
+function wrap_jump_in_integcallback(cache::VRDirectCBEventCache, jump)
+    new_cb = ContinuousCallback(cache, cache;
+        idxs = jump.idxs,
+        rootfind = jump.rootfind,
+        interp_points = jump.interp_points,
+        save_positions = jump.save_positions,
+        abstol = jump.abstol,
+        reltol = jump.reltol)
+    return new_cb
+end
+
+function build_variable_integcallback(cache::VRDirectCBEventCache, cb, jump, jumps...)
+    new_cb = wrap_jump_in_integcallback(cache::VRDirectCBEventCache, jump)
+    build_variable_integcallback(cache, CallbackSet(cb, new_cb), jumps...)
+end
+
+function build_variable_integcallback(cache::VRDirectCBEventCache, cb, jump)
+    CallbackSet(cb, wrap_jump_in_integcallback(cache, jump))
+end
```

Comment:
This seems to be creating one continuous callback per jump, but isn't the whole point of the new method to just create one in total?